### ISSUE TITLE
Core refactor

### ISSUE THREAD

**Issue by robwormald** [ngrx/store#31]
tldr; refactor guts of store to be more compatible with Redux ecosystem. carrying on from #16 

Notes
- store should maintain Rx semantics - Store should continue implement both Observable and Observer, eg `store.subscribe(...)` and `someActionObservable.subscribe(store)`
- drop internal Observables / zip in favor of more redux-like usage - investigate perf implications here. Initial feeling is less subscriptions is better - do we lose anything? 
- drop separate dispatcher injectable? Maintain it but don't require it to be injected?
- use DI to provide both a _root_ reducer and (optional) redux-style enhancers? eg `constructor(rootReducer: RootReducerOpaqueToken, enhancers?: StoreEnhancers)`
- Leverage `Observable.proto.let` for middleware-esque functionality?
- open new repo for store-devtools and a sample monitor?
- favor DI usage over react style flags

Todos:
- investigate Redux Devtools API - can we refactor so both Redux and Store can use common set of tooling?

cc @gaearon @MikeRyan52

**Comment by gaearon** [comment]
_subscribed_, if there’s anything on our end you’d like to happen, please open an issue!

**Comment by MikeRyanDev** [comment]
The difficult challenge is supporting Redux store enhancers that lift `dispatch`, `getState`, or the internal state structure while maintaining Rx semantics.

**Comment by robwormald** [comment]
@MikeRyan52 refactored the guts to be far simpler, see https://github.com/ngrx/store/blob/core-refactor/src/store.ts - want to try devtools stuff against that?

**Comment by MikeRyanDev** [comment]
Yup, will do!

**Comment by MikeRyanDev** [comment]
Ok, finished refactoring devtools on top of `core-refactor`: https://github.com/MikeRyan52/store/tree/feature/devtools-v2

Here's the diff: https://github.com/ngrx/store/compare/core-refactor...MikeRyan52:feature/devtools-v2?expand=1

I had to make some very light changes, but overall it is much easier to lift this new version of Store. 
- The init action has to be `@@redux/init`. The instrumentation code [reduces that action type](https://github.com/gaearon/redux-devtools/blob/master/src/instrument.js#L264) specifically. @gaearon, any thoughts as to how we might be able to generalize this? 
- I had to make `_reducer` public, otherwise TS complained that `LiftedStore` was not assignable to `Store`. Not sure what the best way would be to solve that.
- Changed `createStore` to implement `StoreCreator`

Otherwise, it works!

**Comment by robwormald** [comment]
rad! i pulled it down, works on my end :shipit: 

<img width="734" alt="screen shot 2016-02-09 at 12 51 59 am" src="https://cloud.githubusercontent.com/assets/534930/12911426/68804af4-cec7-11e5-8d44-aac0c1a37c82.png">

A few thoughts:
- Not sure on the naming. I got really tangled up with the term `LiftedStore`, since it was actually the expose store and wrapped a store that had a `lifted` property. Maybe call it "InstrumentedStore"?
- Part of the idea of the `createStore` signature you tweaked was to allow end users to have a nice interface when setting up - totally cool, but I'd like to provide a nicer interface (your new `createStore` method can be internal - the public we could continue to use `provideStore`:

```
bootstrap(App, [ provideStore(reducers, initialState, ...enhancers?) ])`
```

or
`provide(Store, createStore(reducers, initialState, instrument?))`

or something similar to that? Not sure, i'll think about it. _ideally_ i'd like to not have to import angular's `provide` at all internally, but that might just be a pipe dream at this point.

Let's see if @gaearon and co will expose those methods for us (I hear he releases like, 6 new versions a week or something so it shouldn't take too long...) and then we'll merge this in. I have a bit more refactoring to do on the core yet but should have it done in the next couple of days.

One thing I tried to get working but couldn't was using the redux chrome extension - https://github.com/zalmoxisus/redux-devtools-extension - not sure why (or if it's even API compatible) but it makes for a pretty compelling demo. If not, yolo, we'll make our own.

Nice work!

**Comment by MikeRyanDev** [comment]
> Not sure on the naming. I got really tangled up with the term LiftedStore, since it was actually the expose store and wrapped a store that had a lifted property. Maybe call it "InstrumentedStore"?

Agreed, bad holdover from my previous attempts. I've renamed it to `InstrumentedStore` and also cleaned up my `StoreEnhancer` so that it can be composed with other store enhancers.

> Let's see if @gaearon and co will expose those methods for us (I hear he releases like, 6 new versions a week or something so it shouldn't take too long...) and then we'll merge this in. 

Also, if and when it gets merged we will have to handle typings for redux-devtools. Do we need to create a repo for the typings and publish them to the [Typings registry](https://github.com/typings/registry)?

> One thing I tried to get working but couldn't was using the redux chrome extension - https://github.com/zalmoxisus/redux-devtools-extension - not sure why (or if it's even API compatible) but it makes for a pretty compelling demo. If not, yolo, we'll make our own.

Yeah, not API compatible. Instrumentation and monitors are setup to work with redux's createStore contract. _Might_ not be too hard to add support for @ngrx/store in the extension. I'll investigate.

**Comment by JamesHenry** [comment]
Just to highlight something I mentioned on gitter - _if_ it is possible to keep the ng2 provide etc. out of the core of the project, it may well be possible to utilise ngrx/store within an ng1 + TypeScript + RxJS 5 app.

I think that combo will not be uncommon amongst ng1 apps looking to migrate gradually to ng2.

**Comment by robwormald** [comment]
this is merged
